From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Yeh, Feng-Hao" <zzz3x2c1@gmail.com>
Date: Tue, 20 Jun 2023 23:13:31 +0800
Subject: [PATCH] Fix PlayerTeleportEvent not being triggered.

In the Folia, the PlayerTeleportEvent is not triggered when entity.teleportAsync is called.
This patch reinstates the triggering of PlayerTeleportEvent.

diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index deaf4b65b6c9f3dbab6578c25fb78ca84507f70b..e81b4e0cd7a4b859f9491b405fe52811e112dc8c 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -3870,6 +3870,28 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
         // TODO any events that can modify go HERE
 
+        // Add teleport events.
+        CraftEntity craftEntity = getBukkitEntity();
+        Location from = craftEntity.getLocation();
+        // Initialize destination.
+        if (yaw == null) yaw = 0f;
+        if (pitch == null) pitch = 0f;
+        Location to = new Location(destination.getWorld(), pos.x, pos.y, pos.z, yaw, pitch);
+        if (this instanceof ServerPlayer) { // if entity is a player, call PlayerTeleportEvent.
+            // Fire PlayerTeleportEvent.
+            PlayerTeleportEvent tpEvent = new PlayerTeleportEvent((org.bukkit.entity.Player) craftEntity, from, to, cause);
+            Bukkit.getServer().getPluginManager().callEvent(tpEvent);
+            // If event cancelled, abort teleporting.
+            if (tpEvent.isCancelled())
+                return false;
+            // Update destination.
+            to = tpEvent.getTo();
+        }
+        destination = ((CraftWorld) to.getWorld()).getHandle();
+        pos = new Vec3(to.getX(), to.getY(), to.getZ());
+        yaw = to.getYaw();
+        pitch = to.getPitch();
+
         // check for same region
         if (destination == this.level()) {
             Vec3 currPos = this.position();
